### PR TITLE
BUG: to_dense now preserves dtype in SparseArray

### LINKED ISSUE #10648: to_dense does not preserve dtype in SparseArray
This isn't a huge deal, but it seems a little odd:

```python
In [1]: import pandas as pd

In [2]: a = pd.SparseArray([True, False, False, False, True], fill_value=False, dtype=bool)

In [3]: a
Out[3]: 
[True, False, False, False, True]
Fill: False
IntIndex
Indices: array([0, 4], dtype=int32)

In [4]: a.dtype
Out[4]: dtype('bool')

In [5]: d = a.to_dense()

In [6]: d
Out[6]: array([ 1.,  0.,  0.,  0.,  1.])

In [7]: d.dtype
Out[7]: dtype('float64')
```

I would have expected `d` to retain the dtype of `bool`. I can cast down, but I am still wasting 7 bytes per element in the process.

### PR DESCRIPTION
**Pull request by ebolyen**
Also fixes values and get_values.

fixes #10648

### CODE REVIEW

File: pandas/sparse/array.py [hunk 1]
```diff
@@ -266,7 +266,8 @@ def to_dense(self, fill=None):
         # fill the nans
         if fill is None:
             fill = self.fill_value
-        if not np.isnan(fill):
+        # nans can only occur arrays of floating scalars
+        if np.issubdtype(self.dtype, np.inexact) and not np.isnan(fill):
```
Comment (jreback) [inline pandas/sparse/array.py]: a better way to do this is just use ``isnull`` which handles all of these things.

File: pandas/sparse/tests/test_array.py [hunk 2]
```diff
@@ -62,6 +66,10 @@ def test_constructor_copy(self):
         not_copy.sp_values[:3] = 0
         self.assertTrue((self.arr.sp_values[:3] == 0).all())
 
+    def test_constructor_match_dtype(self):
+        res = SparseArray(self.barr_data, dtype=bool)
+        self.assertEqual(res.dtype, bool)
```
Comment (jreback) [inline pandas/sparse/tests/test_array.py]: actually construct what it should look like as an expected and use ``assert_sp_array_equal`` (which might need the same change to use ``isnull`` above)

**Comment by ebolyen** [comment]
@jreback sorry for being slow in getting back to this, I'll make the changes requested before the week is out.

**Comment by jreback** [comment]
can you update

**Comment by jreback** [comment]
@ebolyen can you update according to comments. need a whatsnew note as well.

File: pandas/sparse/tests/test_array.py [hunk 3]
```diff
@@ -81,9 +89,19 @@ def _get_base(values):
         assert(_get_base(arr2) is _get_base(self.arr))
 
     def test_values_asarray(self):
-        assert_almost_equal(self.arr.values, self.arr_data)
-        assert_almost_equal(self.arr.to_dense(), self.arr_data)
-        assert_almost_equal(self.arr.sp_values, np.asarray(self.arr))
+        for arr, arr_data in ((self.arr, self.arr_data),
```
Comment (jreback) [inline pandas/sparse/tests/test_array.py]: include the issue number as a comment here

**Comment by jreback** [comment]
can you update according to comments

**Comment by jreback** [comment]
closing, but pls reopen if you'd like to work on this again.

**Pull request closed without merging**